compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -0,0 +1,409 @@
+import dotty.tools.dotc.typer.Inliner
+
+import scala.annotation.constructorOnly
+
+object Splicing:
+  val name: String = "splicing"
+
+/** Transforms level 1 splices into holes. To do so it transforms the contents of the splice into
+ *  a lambda that receives all cross-quote references.
+ *
+ *  Cross-quote reference is a reference to a definition that is not defined in the current quote.
+ *  Those references appear in quotes that are nested in a splice.
+ *
+ *  After this phase we have the invariant where all splices have the following shape
+ *  ```
+ *  {{{ <holeIdx> | <holeType> | <captures>* | (<capturedTerms>*) => <spliceContent> }}}
+ *  ```
+ *  where `<spliceContent>` does not contain any free references to quoted definitions and `<captures>*`
+ *  contains the quotes with references to all cross-quote references. There are some special rules
+ *  for references in the LHS of assignments and cross-quote method references.
+ *
+ *  In the following code example `x1` and `x2` are cross-quote references.
+ *  ```
+ *  '{ ...
+ *    val x1: T1 = ???
+ *    val x2: T2 = ???
+ *    ${ (q: Quotes) ?=> f('{ g(x1, x2) }) }: T3
+ *  }
+ *  ```
+ *
+ *  This phase identifies cross-quote references such as `x1` and replaces it with an `${x1$}`.
+ *  All cross-quote arguments are directly applied in the lambda.
+ *
+ *  ```
+ *  '{ ...
+ *     val x1: T1 = ???
+ *     val x2: T2 = ???
+ *     {{{ 0 | T3 | x1, x2 |
+ *       (x1$: Expr[T1], x2$: Expr[T2]) => // body of this lambda does not contain references to x1 or x2
+ *         (q: Quotes) ?=> f('{ g(${x1$}, ${x2$}) })
+ *
+ *     }}}
+ *   }
+ *   ```
+ *
+ *  and then performs the same transformation on `'{ g(${x1$}, ${x2$}) }`.
+ *
+ */
+class Splicing extends MacroTransform:
+  import tpd._
+
+  override def phaseName: String = Splicing.name
+
+  override def run(using Context): Unit =
+    if (ctx.compilationUnit.needsQuotePickling) super.run(using freshStagingContext)
+
+  protected def newTransformer(using Context): Transformer = Level0QuoteTransformer
+
+  /** Transforms all quotes at level 0 using the `QuoteTransformer` */
+  private object Level0QuoteTransformer extends Transformer:
+    override def transform(tree: tpd.Tree)(using Context): tpd.Tree =
+      assert(level == 0)
+      tree match
+        case Apply(Select(Apply(TypeApply(fn,_), List(code)),nme.apply),List(quotes))
+        if fn.symbol == defn.QuotedRuntime_exprQuote =>
+          QuoteTransformer().transform(tree)
+        case TypeApply(_, _) if tree.symbol == defn.QuotedTypeModule_of =>
+          QuoteTransformer().transform(tree)
+        case tree: DefDef if tree.symbol.is(Inline) =>
+          tree
+        case _ =>
+          super.transform(tree)
+  end Level0QuoteTransformer
+
+
+  /** Transforms all direct splices in the current quote and replace them with holes. */
+  private class QuoteTransformer() extends Transformer:
+    private val quotedDefs = mutable.Set.empty[Symbol]
```

Comment:
Good example.

First, the `Level0QuoteTransformer` finds that quote. Then the `QuoteTransformer` will transform the contents of the quote collecting the `quotedDefs`. When it reaches the splice

```
${
         // ...
         '{
             // ...
             ${ f('a) }
          }
     }
```
`quotedDefs = Set(<val a>)`.

Then the `SpliceTransformer` will transform the contents of the splice and create the hole. This is done with a snapshot of the current `quotedDefs` (the new version of the code is more explicit about this). The `SpliceTransformer` will only transform the captured definition reference and will just transform within quotes.
After this transformation, we get a hole like
```
    {{{ 0 | T | a |
      (x1: Expr[T2]) =>
         // ...
         '{
             // ...
             ${ f('{${x1}}) }
          }
    }}}
```
Then we recursively transform the contents of the hole using `Level0QuoteTransformer` which will find any quote can transform it. In this case, it will find
```
'{
       // ...
      ${ f('{${x1}}) }
}
```

Note that `QuoteTransformer` will also cancel splices. Therefore it will transform `'{${x1}}` into `x1`. Removing unnecessary splices that might have been added in the process.

Reply:
Thanks for the detailed explanation. If I understand correctly, the current algorithm will be quadratic in terms of the tree size if there are such nested quote/splices (which should be rare).

A digression: I am wondering if a linear algorithm is possible. As such use cases are rare, there is no need for premature optimization in this PR.

Reply:
Linear might be possible but it would be much harder to reason about. 

There is also the possibility of detecting if the contents of a splice need further transformation while doing the first pass. Then avoid the second pass if not needed.

Another alternative is to warn the user where there are to many nested quotes or code in a splice is too large. The contents of a splice can always be extracted into a separate method. This is what we recommend in the best practices for legibility.